srtcore/group.cpp

```diff
@@ -1062,8 +1062,16 @@ void CUDTGroup::send_CheckValidSockets()
         CUDTSocket* revps = m_pGlobal->locateSocket_LOCKED(d->id);
         if (revps != d->ps)
         {
-            HLOGC(gmlog.Debug, log << "group/send_CheckValidSockets: socket @" << d->id << " is no longer valid, REMOVING FROM $" << id());
-            remove_LOCKED(d->id);
+            // Note: the socket might STILL EXIST, just in the trash, so
+            // it can't be found by locateSocket. But it can still be bound
+            // to the group. Just mark it broken from upside so that the
+            // internal sending procedures will skip it. Removal from the
+            // group will happen in GC, which will both remove from
+            // group container and cut backward links to the group.
+
+            HLOGC(gmlog.Debug, log << "group/send_CheckValidSockets: socket @" << d->id << " is no longer valid, setting BROKEN in $" << id());
+            d->sndstate = SRT_GST_BROKEN;
+            d->rcvstate = SRT_GST_BROKEN;
```

Comment:
Are we safe to set this to broken? This happens under these locks:
- `m_pGlobal->m_GlobControlLock`
- `m_GroupLock`

Reply:
Correct. The first one protects the sockets against being updated against the group (that is, removed from the group container in the meantime), the second one protects the `m_Group` container (members container). Note also that this function is being called from inside the sending function only, so it happens in the same thread as the transmission functions.  A simultaneous call to a receiving function, let's say, isn't possible because it also requires lock on m_Group.